slasher/detection/attestations/detect_attestation.go

```diff
@@ -0,0 +1,123 @@
+package attestations
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+)
+
+// DetectAttestationForSlashings returns an attester slashing if the attestation submitted
+// is a slashable vote.
+func (d *AttDetector) DetectAttestationForSlashings(ctx context.Context, req *ethpb.IndexedAttestation) ([]*ethpb.AttesterSlashing, error) {
+	//TODO(#3133): add signature validation
+	if req.Data == nil {
+		return nil, fmt.Errorf("cant hash nil data in indexed attestation")
+	}
+	indices := req.AttestingIndices
+	root, err := hashutil.HashProto(req.Data)
+	if err != nil {
+		return nil, err
+	}
+	var attSlashingResp []*ethpb.AttesterSlashing
+	attSlashings := make(chan []*ethpb.AttesterSlashing, len(indices))
+	errorChans := make(chan error, len(indices))
+	var wg sync.WaitGroup
+	lastIdx := int64(-1)
+	for _, idx := range indices {
+		if int64(idx) <= lastIdx {
+			return nil, fmt.Errorf("indexed attestation contains repeated or non sorted ids")
+		}
+		wg.Add(1)
+		go func(idx uint64, root [32]byte, req *ethpb.IndexedAttestation) {
+			atts, err := d.DoubleVotes(idx, root[:], req)
+			if err != nil {
+				errorChans <- err
+				wg.Done()
+				return
+			}
+			if atts != nil && len(atts) > 0 {
+				attSlashings <- atts
+			}
+			atts, err = d.DetectSurroundVotes(ctx, idx, req)
+			if err != nil {
+				errorChans <- err
+				wg.Done()
+				return
+			}
+			if atts != nil && len(atts) > 0 {
+				attSlashings <- atts
+			}
+			wg.Done()
+			return
+		}(idx, root, req)
+	}
+	wg.Wait()
+	close(errorChans)
+	close(attSlashings)
+	for e := range errorChans {
+		if err != nil {
```

Comment:
What is going on here?

Reply:
errors that have been collected in the channel are concatenated for output